lib/stitch_plan/lock_stitch.py

```diff
@@ -0,0 +1,202 @@
+        if pos == 'end':
+            path = reversed(path)
+
+        lock_stitches = []
+        for i, stitch in enumerate(path):
+            stitch = Stitch(stitch[0], stitch[1], tags=('lock_stitch',))
+            lock_stitches.append(stitch)
+        return lock_stitches
+
+    def _absolute_tie(self, stitches, pos):
+        # make sure the path consists of only floats
+        path = string_to_floats(self.path, " ")
+
+        # get the length of our lock stitch path
+        if pos == 'start':
+            lock_pos = []
+            lock = 0
+            # reverse the list to make sure we end with the first stitch of the target path
+            for tie_path in reversed(path):
+                lock = lock - tie_path * self.scale_absolute
+                lock_pos.insert(0, lock)
+        if pos == 'end':
+            lock_pos = []
+            lock = 0
+            for tie_path in path:
+                lock = lock + tie_path * self.scale_absolute
+                lock_pos.append(lock)
+        max_lock_length = max(lock_pos)
+
+        # calculate the amount stitches we need from the target path
+        # and generate a line
+        upcoming = [stitches[0]]
+        for stitch in stitches[1:]:
+            to_start = stitch - upcoming[-1]
+            upcoming.append(stitch)
+            if to_start.length() >= max_lock_length:
+                break
+        line = LineString(upcoming)
+
+        # add tie stitches
+        lock_stitches = []
+        for i, tie_path in enumerate(lock_pos):
+            if tie_path < 0:
+                stitch = Stitch(stitches[0] + tie_path * (stitches[1] - stitches[0]).unit())
+            else:
+                stitch = substring(line, start_dist=tie_path, end_dist=tie_path)
+                stitch = Stitch(stitch.x, stitch.y, tags=('lock_stitch',))
+            lock_stitches.append(stitch)
+        return lock_stitches
+
+    def _relative_tie(self, stitches, pos):
+        path = string_to_floats(self.path, " ")
+
+        to_previous = stitches[1] - stitches[0]
+        length = to_previous.length()
+
+        lock_stitches = []
+        if length > 0.5 * PIXELS_PER_MM:
+
+            # Travel back one stitch, stopping halfway there.
+            # Then go forward one stitch, stopping halfway between
+            # again.
+
+            # but travel at most 1.5 mm
+            length = min(length, 1.5 * PIXELS_PER_MM)
+
+            direction = to_previous.unit()
+
+            for delta in path:
+                lock_stitches.append(Stitch(stitches[0] + delta * length * direction, tags=('lock_stitch')))
+        else:
+            # Too short to travel part of the way to the previous stitch; just go
+            # back and forth to it a couple times.
+            for i in (1, 0, 1, 0):
+                lock_stitches.append(stitches[i])
+        return lock_stitches
+
+
+def get_lock_stitch_by_id(pos, lock_type, default="half_stitch"):
+    id_list = [str(lock) for lock in LOCK_DEFAULTS[pos]]
```

Comment:
Why `str(lock)` instead of `lock.id`?

Reply:
Not sure why I did that. Maybe due to the way it used to work before the rewrite. I changed it.